### PR TITLE
API: Add SessionCatalog interface and base class

### PR DESCRIPTION
**Pull request by rdblue**
This adds a `SessionCatalog` interface for engines that share catalog instances across sessions.

`SessionCatalog` is like `Catalog`, but passes a `SessionContext` object in each call. This session context can be used to carry identity and credentials, which are passed through to the catalog service. `SessionCatalog` does not extend `Catalog` to avoid bugs where context is accidentally not passed.

Because `SessionCatalog` does not implement `Catalog`, a base implementation class is also included in this PR that adapts `SessionCatalog` to `Catalog` using `withContext` and `asCatalog` methods. For example:

```java
  boolean exists = sessionCatalog.withContext(sessionContext, (Catalog catalog) -> {
    return catalog.tableExists(tableIdent);
  });
```

This was originally part of #4575, but is not a separate PR to avoid blocking other work.

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/catalog/SessionCatalog.java [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+   *
+   * @param context session context
+   * @param ident a table identifier
+   * @param schema a schema
+   * @return the builder to create a table or start a create/replace transaction
+   */
+  Catalog.TableBuilder buildTable(SessionContext context, TableIdentifier ident, Schema schema);
+
+  /**
+   * Register a table with if it does not exist.
```
Comment (danielcweeks) [inline api/src/main/java/org/apache/iceberg/catalog/SessionCatalog.java:144]: ~~with~~

File: core/src/main/java/org/apache/iceberg/catalog/BaseSessionCatalog.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+  public Catalog asCatalog(SessionContext context) {
+    return catalogs.get(context.sessionId(), id -> new AsCatalog(context));
+  }
+
+  public <T> T withContext(SessionContext context, Function<Catalog, T> task) {
+    return task.apply(asCatalog(context));
+  }
+
+  public class AsCatalog implements Catalog, SupportsNamespaces {
```
Comment (danielcweeks) [inline core/src/main/java/org/apache/iceberg/catalog/BaseSessionCatalog.java:67]: Can this inner class be static?  Caches historically have reference problems with non-static inner classes (though I don't recall the exact scenarios).

Comment (danielcweeks) [reply core/src/main/java/org/apache/iceberg/catalog/BaseSessionCatalog.java:67]: I don't think this usage has the issues I was concerned about. Resolving.

**Review by danielcweeks — APPROVED** [review]
Very minor nit on comment text.  +1

**Comment by rdblue** [comment]
Thanks, @danielcweeks! And to everyone that looked at this in #4575.

**Pull request merged**